File: examples/15-cli-commands/main.go
// examples/15-cli-commands/main.go builds a reference presentation documenting
// the gopptx CLI subcommands.
//
// No CLI execution is performed — this example creates a PPTX that serves as
// a quick-reference guide for the available commands and their purpose.
//
// Run with: go run ./examples/15-cli-commands/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	log "github.com/djinn-soul/gopptx/pkg/stdlog"
)

const (
	outputDir  = "examples/output"
	outputFile = "15_cli_commands.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output directory: %w", err)
	}

	slides := []pptx.SlideContent{
		// Slide 1: title / overview
		pptx.NewSlide("gopptx CLI Reference").
			AddBullet("The gopptx binary exposes several subcommands for common PPTX tasks.").
			AddBullet("Each subsequent slide documents one subcommand.").
			AddBullet("Run `gopptx --help` to see all available commands."),

		// Slide 2: create
		pptx.NewSlide("gopptx create").
			AddBullet("Creates a new blank PPTX presentation.").
			AddBullet("Usage:  gopptx create --title \"My Deck\" --slides 5 -o out.pptx").
			AddBullet("Flags:").
			AddBullet("  --title   Presentation title (required)").
			AddBullet("  --slides  Number of blank slides to generate").
			AddBullet("  -o        Output file path"),

		// Slide 3: md2ppt
		pptx.NewSlide("gopptx md2ppt").
			AddBullet("Converts a Markdown file into a PPTX presentation.").
			AddBullet("Usage:  gopptx md2ppt input.md -o out.pptx").
			AddBullet("Each top-level heading becomes a slide title.").
			AddBullet("Bullet lists under a heading become slide bullets.").
			AddBullet("Code blocks are rendered as monospace text boxes."),

		// Slide 4: info
		pptx.NewSlide("gopptx info").
			AddBullet("Displays metadata and structural information about a PPTX file.").
			AddBullet("Usage:  gopptx info presentation.pptx").
			AddBullet("Output includes:").
			AddBullet("  - Title, author, and creation date").
			AddBullet("  - Slide count and slide sizes").
			AddBullet("  - Embedded media and chart counts"),

		// Slide 5: validate
		pptx.NewSlide("gopptx validate").
			AddBullet("Validates the structural integrity of a PPTX file.").
			AddBullet("Usage:  gopptx validate presentation.pptx").
			AddBullet("Checks for:").
			AddBullet("  - Missing or malformed XML parts").
			AddBullet("  - Broken relationship references").
			AddBullet("  - Invalid media or chart entries").
			AddBullet("Exits with a non-zero code if issues are found."),

		// Slide 6: merge
		pptx.NewSlide("gopptx merge").
			AddBullet("Merges two or more PPTX files into a single presentation.").
			AddBullet("Usage:  gopptx merge a.pptx b.pptx -o merged.pptx").
			AddBullet("Slides are appended in the order the input files are given.").
			AddBullet("Themes and masters from the first file are preserved.").
			AddBullet("Assets (images, charts) are deduplicated automatically."),

		// Slide 7: version
		pptx.NewSlide("gopptx version").
			AddBullet("Prints the current gopptx version and build information.").
			AddBullet("Usage:  gopptx version").
			AddBullet("Output example:").
			AddBullet("  gopptx v1.2.3 (commit abc1234, built 2025-01-01)").
			AddBullet("Useful for confirming the installed binary in CI pipelines."),
	}

	outputPath := filepath.Join(outputDir, outputFile)
	if err := pptx.WriteFile(outputPath, "gopptx CLI Reference", slides); err != nil {
		return fmt.Errorf("write presentation: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

File: examples/15-cli-commands/main.py
"""Build a quick-reference guide for the gopptx CLI subcommands.

This example demonstrates:
- Creating a multi-slide reference presentation in code
- Using add_bullet_slide() to document CLI commands
- No CLI execution is performed — this generates a PPTX reference guide
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation


def _add_overview_slide(prs: Presentation) -> None:
    """Add a title and overview slide."""
    prs.add_bullet_slide(
        "gopptx CLI Reference",
        [
            "The gopptx binary exposes several subcommands for common PPTX tasks.",
            "Each subsequent slide documents one subcommand.",
            "Run `gopptx --help` to see all available commands.",
        ],
    )


def _add_create_slide(prs: Presentation) -> None:
    """Add a slide documenting the 'create' subcommand."""
    prs.add_bullet_slide(
        "gopptx create",
        [
            "Creates a new blank PPTX presentation.",
            'Usage:  gopptx create --title "My Deck" --slides 5 -o out.pptx',
            "Flags:",
            "  --title   Presentation title (required)",
            "  --slides  Number of blank slides to generate",
            "  -o        Output file path",
        ],
    )


def _add_md2ppt_slide(prs: Presentation) -> None:
    """Add a slide documenting the 'md2ppt' subcommand."""
    prs.add_bullet_slide(
        "gopptx md2ppt",
        [
            "Converts a Markdown file into a PPTX presentation.",
            "Usage:  gopptx md2ppt input.md -o out.pptx",
            "Each top-level heading becomes a slide title.",
            "Bullet lists under a heading become slide bullets.",
            "Code blocks are rendered as monospace text boxes.",
        ],
    )


def _add_info_slide(prs: Presentation) -> None:
    """Add a slide documenting the 'info' subcommand."""
    prs.add_bullet_slide(
        "gopptx info",
        [
            "Displays metadata and structural information about a PPTX file.",
            "Usage:  gopptx info presentation.pptx",
            "Output includes:",
            "  - Title, author, and creation date",
            "  - Slide count and slide sizes",
            "  - Embedded media and chart counts",
        ],
    )


def _add_validate_slide(prs: Presentation) -> None:
    """Add a slide documenting the 'validate' subcommand."""
    prs.add_bullet_slide(
        "gopptx validate",
        [
            "Validates the structural integrity of a PPTX file.",
            "Usage:  gopptx validate presentation.pptx",
            "Checks for:",
            "  - Missing or malformed XML parts",
            "  - Broken relationship references",
            "  - Invalid media or chart entries",
            "Exits with a non-zero code if issues are found.",
        ],
    )


def _add_merge_slide(prs: Presentation) -> None:
    """Add a slide documenting the 'merge' subcommand."""
    prs.add_bullet_slide(
        "gopptx merge",
        [
            "Merges two or more PPTX files into a single presentation.",
            "Usage:  gopptx merge a.pptx b.pptx -o merged.pptx",
            "Slides are appended in the order the input files are given.",
            "Themes and masters from the first file are preserved.",
            "Assets (images, charts) are deduplicated automatically.",
        ],
    )


def _add_version_slide(prs: Presentation) -> None:
    """Add a slide documenting the 'version' subcommand."""
    prs.add_bullet_slide(
        "gopptx version",
        [
            "Prints the current gopptx version and build information.",
            "Usage:  gopptx version",
            "Output example:",
            "  gopptx v1.2.3 (commit abc1234, built 2025-01-01)",
            "Useful for confirming the installed binary in CI pipelines.",
        ],
    )


def main() -> None:
    """Create a presentation documenting the gopptx CLI subcommands."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    with Presentation.new("gopptx CLI Reference") as prs:
        _add_overview_slide(prs)
        _add_create_slide(prs)
        _add_md2ppt_slide(prs)
        _add_info_slide(prs)
        _add_validate_slide(prs)
        _add_merge_slide(prs)
        _add_version_slide(prs)

        output_path = output_dir / "15-cli-commands.pptx"
        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("Created a 7-slide CLI reference guide covering:")
    print("  create, md2ppt, info, validate, merge, version")


if __name__ == "__main__":
    main()
